test/unit/incoming-message-stream-test.js

```diff
@@ -175,4 +176,49 @@ describe('IncomingMessageStream', function() {
       incoming.resume();
     });
   });
+
+  it('should validate packet header size', function(done) {
+    const packetData = Buffer.from('test1234');
+    const packetHeader = Buffer.alloc(7);
+
+    let offset = 0;
+    offset = packetHeader.writeUInt8(0x11, offset);
+    offset = packetHeader.writeUInt8(0x01, offset);
+    offset = packetHeader.writeUInt16BE(8 + packetData.length, offset);
+    offset = packetHeader.writeUInt16BE(0x0000, offset);
+    packetHeader.writeUInt8(0x00, offset);
+
+    const packet = Buffer.concat([packetHeader, packetData]);
+
+    const incoming = new IncomingMessageStream(new Debug());
+
+    assert.throws(() => {
+      incoming.end(packet);
+    }, ConnectionError, 'Unable to process incoming packet');
+
+    done();
```

Comment:
So, this test "works", but there are some slight issues with it.

For one, we want to test whether an exception is emitted via the `error` event, because that is what happens if the `callback` is called with an error.

The reason why catching the error via `assert.throws` works here is because this ends up being an uncaught exception. When the packet is parsed, the callback is called with the error, but because there is no handler on the `error` event, the error is being thrown, which is caught by `assert.throws`, and the test passes.

If you'd change the `callback` invocation to `throw` directly inside `IncomingMessageStream`, the test would still pass, but if that condition would be hit in a real application, the application would still break due to an uncaught exception! So the test is giving us a false sense of confidence into our code being "correct".

What would be better is if we'd test that the `error` event is emitted, like this:

```suggestion
    incoming.on('error', (err) => {
      assert.instanceOf(err, ConnectionError);
      assert.equal(err.message, 'Unable to process incoming packet');

      done();
    });

    incoming.end(packet);
```

Reply:
I did try incoming.on('error'....) but the test gets hanged and the event 'error' doesn't get caught or isn't being emitted.